### ISSUE TITLE
Processing lint results (placing signs) takes a long time on large files

### ISSUE THREAD

**Issue by benekastah** [benekastah/neomake#48]
I don't know exactly what is slow, so I'll just write what I did so that it will be easy to reproduce later:

```
let g:neomake_c_nvimlint_maker = {
    \ 'exe': './clint.py',
    \ 'errorformat': '%f:%l:  %m [%n]'
    \ }
```

Then edit [eval.c](https://github.com/neovim/neovim/blob/master/src/nvim/eval.c) and run `:Neomake nvimlint`. It gave over 900 results for me and the file is close to 20000 loc. It seems like when I disabled placing of signs with `let g:neomake_place_signs = 0` that it helped a lot. I'm not sure if that's because the file is too large or if my sign placing routine is just too slow for 900 entries.

One idea: Only process the results that are correspond with the visible part of the buffer.

**Comment by benekastah** [comment]
Another idea that may speed things up is to place all the signs as they come, but only move them to the top after the maker is complete. I suspect the moving signs to the top part is the expensive part.

**Comment by Neki** [comment]
Even on small files (~ 100 lines), processing results when there is more than 10 errors has a noticeable delay (between half a second and a second) on my computer (neovim is freezed during the processing). I have started to play with this issue, so far it seems the best gain I was able to achieve was by removing the `redraw!` command at https://github.com/benekastah/neomake/blob/master/autoload/neomake.vim#L338 (but performance is still poor on `eval.c`). What is the goal of this command? Signs appear as normal when it is removed. 

Removing the "move signs on top" routine did not speed things up noticeably.

Then again I did not benchmark the improvements or anything, so take this with a grain of salt.

**Comment by benekastah** [comment]
It's possible that the `redraw!` is no longer needed with recent versions of neovim. When I coded that stuff up at first I would get odd results without it.

There are a lot of issues with the sign placing code (as you can see by the issues list). I'm planning to refactor that stuff this weekend where I hope to knock out a bunch of the major issues. So keep that in mind as you investigate that code and don't sink too much time into it because it will change. I'll try to get that done quickly so that it doesn't block you on looking into issues when you want to.

**Comment by Neki** [comment]
Ok, thanks for the info. There is no need to rush as far as I am concerned, I have plenty of other things to do :)

**Comment by benekastah** [comment]
I said I fixed this in #77, but I'd like to test it again before closing.

**Comment by Neki** [comment]
Some tests on my laptop:

On "regular" files (with a sane number of lines) the situation seems to have improved a bit, there are still some micro freezes when results are processed but it is completely acceptable. Thanks!

Neovim still freezes on `eval.c` (for 15 seconds or so), before the signs are even placed. So maybe the problem now is not placing signs, but just parsing the output of the maker.

In both cases scrolling is completely smooth.

**Comment by benekastah** [comment]
Hmm. Ok, I'll see what I can do about that 15 second wait. That's pretty bad.

**Comment by benekastah** [comment]
This is much improved in the [performance branch](https://github.com/benekastah/neomake/tree/performance). I just want to test that branch at work tomorrow to make sure I didn't break anything. I got two back to back runs of neomake on eval.c from 39 seconds to just over 8 seconds. During that 8 seconds, the editor is pretty usable. Here's the maker I used:

```
let g:neomake_c_clintpy_maker = {
    \ 'exe': './clint.py',
    \ 'errorformat': '%f:%l:  %m [%n]',
    \ }
```
